### ISSUE TITLE
[Feature Request] Add the possibility to execute after FBT runs

### ISSUE THREAD

**Issue by saitoh183** [Laharah/deluge-FileBotTool#10]
Would it be possible to add the possibility to execute a script/cmd after FBT does it thing? Sort of how the AMC script from FB has --def exec command

**Comment by Laharah** [comment]
Next time I add features, I'll look into adding this one in.

**Comment by Laharah** [comment]
I see that an `--exec` command [has been added to FileBot](https://www.filebot.net/forums/viewtopic.php?f=4&t=5047). I'll see if there's a good way to implement this into the plugin.

**Comment by rjsachse** [comment]
I was thinking more like the execute plugin not using filebot -exec. Once filebot has finished then execute a script. But i guess it might be hard to know when filebot has finished.

**Comment by rjsachse** [comment]
Also would be great to pass renamed file name arg to script like execute plugin.

**Comment by Laharah** [comment]
well, it looks like the `--exec` flag uses the same argument template as the other [format expressions](http://www.filebot.net/naming.html). If that's so, you should be able to have an argument like `scp {f} server:{f}`. 

The last time I looked into this, before the exec flag was implemented, I tried to make the plugin play nice with the execute plugin that comes with deluge. Unfortunately, the execute plugin listens for the same `torrent_finished` event that filebottool does, so it will run at the same time/before filebot can sort anything. To stop that from happening, the plugin would have to insert a shim into the execute plugin and monkey patch it on the fly. However there's no way to guarantee that filebottool is loaded *after* the execute plugin, so you'd have to do something like temporary patching and un-patching. It all gets super messy, and it's generally considered bad behavior to monkey patch someone else's plugin anyways.

As I see it. There are really just 2 options:

1. Add support for the `--exec` option built into filebot. Easy to implement, a little difficult to integrate from a clean UI perspective

2. Write a second plugin called something like **FileBotTool-Execute** that basically just replicates the Execute plugin, but triggers based on a FileBotTool custom event. It would be fairly straight forward to pass the handler info in the event to support multiple actions for different handlers.

Both solutions would requre a good amount of thought about special cases:

* Should the execute commands run every time or only on auto-executed renames?
* What should happen if you run a rename multiple times? What if you revert a rename?
* How can you make it easy to turn off the commands if you're doing a one-off rename?
* Should the commands be saved with the handler, or should they have their own auto-execute rules?

I'll try to think about it over the next few days. I'm open to suggestions, opinions or pull requests :)

**Comment by rjsachse** [comment]
You make some very good points.
I think commands should be saved with the handler. I think really haveing the commands execute with auto-rename, if you are doing it manually how hard is it to just run a command after.

My setup is I add a torrent and lable plus check rules to apply the right lable, then when finished filebottool rebames and puts in right location. Now i have to manually update library in kodi. I have written a simple bash script that sends a notification to user and then commands kodi to update library.

Think commands really are for auto-move/rename

I see there is a updated execute plugin that might be good to use instead of creating a new plugin

https://github.com/ultnrg/deluge-execute/blob/master/README.md

I haven't done any deluge plugin yet only just moved from windows to linux and only just started learning python but do have some experience in other languages so sort of understand the code.

So maybe the easiest way would be to add filebotevent in the updated execute plugin.

**Comment by rjsachse** [comment]
Well after some research and trial and error trying to compile latest filebot, looks like this would only work on mac and windows as filebot 4.8 would be needed to use the -exec. I am on linux debian stretch on a arm device (rpi3). So you need to check filebot version and make sure it is above 4.7.9 for it to work that way.

**Comment by Laharah** [comment]
So, I gather that most of the users of my plugin use the linux version of 4.7.9 (myself included) and despite donating to the (patreon)[https://www.patreon.com/filebot/] it doesn't look like linux support will be coming very soon unless more people sign up. I think an additional plugin may be the way to go with this. It will require a bit more effort on the side of users to find and add an additional plugin, but I figure that most people looking to automate will be savvy enough to figure it out.

Unfortunately I've been a bit swamped with real life work. So aside from fixing issues I personally encounter with my plugin, support has been a bit lacking lately. I'll try to pencil in some time to hammer out at least what's needed in *this* plugin to support it. Namely a `filebot_rename_finished` event that get's triggered after a rename. 

Once that's done the work of making the new execute plugin can start. I think most of the work can be mitigated by simply starting with the deluge execute plugin and modifying it. Changing one line should be enough to get it to trigger on the new event, then it's simply a matter of adding UI and maybe some parsing and passing of data that comes with the event.

**Comment by rjsachse** [comment]
Thanks. Yes life has been busy. I will try and have a look at this soon and trying to find time aswell.

**Comment by rjsachse** [comment]
Hello again. I have made some adjustments to the deluge execute plugin. I am just testing now. Nothing special just should pick up the events. If works i will upload to github.

**Comment by Laharah** [comment]
@rjsachse cool, let me know if you need any help. Keep in mind you'll have to clone and compile [FileBotTool 1.1.15](https://github.com/Laharah/deluge-FileBotTool/tree/1.1.15) (`git checkout 1.1.15`) manually since the events are only emitted in the develop branch right now. They won't be in an actual release until I land them in master when I finish with 1.2.

**Comment by rjsachse** [comment]
Yes i already compile 1.1.15. I just cloned deluge execute plugin and have attempted to implement the events. I have not tried it yet.

https://github.com/rjsachse/deluge-FBTExecute

**Comment by rjsachse** [comment]
ok i have a bit of time this morning and gave it ago bu i get this error

`[ERROR   ] 08:03:14 eventmanager:59 Event handler FileBotToolTorrentFinishedEvent failed in <function event_handler at 0x6c139a30> with exception event_handler() takes exactly 1 argument (2 given)`

**Comment by Laharah** [comment]
It looks like you're trying to catch the event with a function that doesn't have the right number of arguments. 

`FileBotToolTorrentFinishedEvent has 2 arguments, so the function that catches them should have arguments for both of those pieces of data for example

```python

class Core(CorePluginBase):

    def enable(self)
        event_manager = component.get("EventManager")
        event_manager.register_event_handler("FileBotToolTorrentFinishedEvent",
                                               self.catch_fb_torrent_finished)
    
    def catch_fb_torrent_finished(self, torrent_id, handler_name):
        log.debug("Caught FileBotToolTorrentFinishedEvent! for torrent: {0}".format(torrent_id))

```

Novice how `catch_fb_torrent_finished has the same arguments as `FileBotToolTorrentFinishedEvent`: https://github.com/Laharah/deluge-FileBotTool/blob/b3b097c188450e3188e0dfc78bcd8a603a7a5862/filebottool/events.py#L12
